### PR TITLE
feat(frontend): support `explain format json`

### PR DESCRIPTION
**Pull request by kwannoel**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Closes https://github.com/risingwavelabs/risingwave/issues/18816 https://github.com/risingwavelabs/risingwave/issues/17218 https://github.com/risingwavelabs/risingwave/issues/14168

We just use the `Pretty` IR, wrap it in a newtype, derive a serde instance for it.

We intentionally do NOT use serde already implemented for the proto output of the to_proto phase (i.e. when `to_stream_prost / to_batch_prost` are called). That's because it locks the plan type to the `stream_proto / batch_proto` type.

EXPLAIN FORMAT JSON should only change the format, and not determine the underlying plan type (logical / stream / batch / to_proto).

Explain terse doesn't need to be supported. Using the json output, we can just do (see before/after below):
```
 jq 'def extract: {name: .name, children: [.children[] | extract]}; extract' plan.json
```

<details>
<summary>before</summary>

```json
    {
      "name": "LogicalProject",
      "fields": {
        "exprs": [
          "approx_percentile($expr1)"
        ]
      },
      "children": [
        {
          "name": "LogicalAgg",
          "fields": {
            "aggs": [
              "approx_percentile($expr1)"
            ]
          },
          "children": [
            {
              "name": "LogicalProject",
              "fields": {
                "exprs": [
                  "t.v1::Float64 as $expr1"
                ]
              },
              "children": [
                {
                  "name": "LogicalScan",
                  "fields": {
                    "columns": [
                      "t.v1",
                      "t._row_id"
                    ],
                    "table": "t"
                  },
                  "children": []
                }
              ]
            }
          ]
        }
      ]
    }
```

</details>

<details>

<summary>after</summary>

```json
{
  "name": "LogicalProject",
  "children": [
    {
      "name": "LogicalAgg",
      "children": [
        {
          "name": "LogicalProject",
          "children": [
            {
              "name": "LogicalScan",
              "children": []
            }
          ]
        }
      ]
    }
  ]
}

```

</details>

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

**Public Preview**: Do note that the structure and format of EXPLAIN FORMAT JSON is considered unstable, and may change overtime.

After this PR, a new [`explain option`](https://docs.risingwave.com/docs/current/sql-explain/#explain-options) is supported
```
FORMAT (TEXT | JSON)
```

For `FORMAT JSON`, it will output json for `logical` and `physical` plans, while it ignores `trace, distsql` options.
For `FORMAT TEXT`, the output is the same as the current output format of explain.

Here's an example:

```sql
create table t(v1 int);
explain(physical, format json) create materialized view m1 as select * from t;
```

```json
  {
   "name": "StreamMaterialize",
   "fields": {
     "columns": [
       "v1",
       "t._row_id(hidden)"
     ],
     "pk_columns": [
       "t._row_id"
     ],
     "pk_conflict": "NoCheck",
     "stream_key": [
       "t._row_id"
     ]
   },
   "children": [
     {
       "name": "StreamTableScan",
       "fields": {
         "columns": [
           "v1",
           "_row_id"
         ],
         "table": "t"
       },
       "children": []
     }
   ]
 }

```

### CODE REVIEW

**Comment by xxchan** [comment]
QQ: Is this just for debugging and not ready to be used widely by users? I assume the format is not finalized and will be subject to change. (If so we can happily merge and try it quickly)

File: src/frontend/src/handler/explain.rs [hunk 1]
```diff
@@ -212,7 +213,10 @@ async fn do_handle_explain(
             ExplainType::Physical => {
                 // if explain trace is on, the plan has been in the rows
                 if !explain_trace && let Ok(plan) = &plan {
-                    blocks.push(plan.explain_to_string());
+                    match explain_format {
+                        ExplainFormat::Text => blocks.push(plan.explain_to_string()),
+                        ExplainFormat::Json => blocks.push(plan.explain_to_json()),
+                    }
```
Comment (xxchan) [inline src/frontend/src/handler/explain.rs:219]: Only physical plan is supported?

File: src/frontend/planner_test/src/lib.rs
Comment (xxchan) [inline src/frontend/planner_test/src/lib.rs:1]: I guess we can directly use `EXPLAIN` in test case, and no need to add them all here

https://github.com/risingwavelabs/risingwave/blob/main/src/frontend/planner_test/README.md#explain-as-the-test-case

File: src/frontend/src/handler/explain.rs (hunk 1 above)
Comment (xxchan) [reply src/frontend/src/handler/explain.rs:219]: From test cases it seems they are supported. The code diff looks strange though

Comment (xxchan) [reply src/frontend/src/handler/explain.rs:219]: BTW, how will `explain (trace)` look like?

**Comment by kwannoel** [comment]
> QQ: Is this just for debugging and not ready to be used widely by users? I assume the format is not finalized and will be subject to change. (If so we can happily merge and try it quickly)

Yeah it's mainly for us to get a structured output of the users query plan.

File: src/frontend/src/utils/pretty_serde.rs [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! @kwannoel:
+//! This module implements Serde for the Pretty struct. Why not implement it directly on our plan nodes?
+//! That's because Pretty already summarizes the fields that are important to us.
+//! You can see that when `explain()` is called, we directly return the `Pretty` struct.
+//! The _proper way_ to do this would be to create a new data structure that plan nodes get converted into,
```
Comment (xxchan) [inline src/frontend/src/utils/pretty_serde.rs:19]: You mean:

```
`Plan` -> `PlanForExplain` -> `Pretty` -> `String`
                          |-> `Json`
```

Interesting

Comment (xxchan) [reply src/frontend/src/utils/pretty_serde.rs:19]: I agree `Pretty` is not targeted as the real "PlanForExplain" IR, but good enough for now 👍

**Review by xxchan — COMMENTED** [review]
Generally LGTM 

(But did not read / play with the format. If it's not finalized, I think we can play with it and refine it as needed

**Comment by xxchan** [comment]
Do we want to document it now? If yes, we should note format stability

File: src/frontend/src/handler/explain.rs (hunk 1 above)
Comment (kwannoel) [reply src/frontend/src/handler/explain.rs:219]: It will ignore the `JSON FORMAT`, and just output text.

Comment (xxchan) [reply src/frontend/src/handler/explain.rs:219]: Hmmm, should we reject it instead?

File: src/frontend/planner_test/src/lib.rs
Comment (st1page) [reply src/frontend/planner_test/src/lib.rs:1]: The name is a little confused. maybe `direct_output` is better

Comment (kwannoel) [reply src/frontend/planner_test/src/lib.rs:1]: Fixed

File: src/frontend/src/handler/explain.rs (hunk 1 above)
Comment (kwannoel) [reply src/frontend/src/handler/explain.rs:219]: Fixed

File: e2e_test/batch/explain.slt [hunk 3]
```diff
@@ -11,4 +11,10 @@ statement ok
...
 statement error Not supported: EXPLAIN CREATE VIEW
-explain create view v as select 1;
\ No newline at end of file
+explain create view v as select 1;
+
+statement error
+explain (trace, format json) select 1;
+
+statement error
+explain (distsql, format json) select 1;
```
[the remark this answers is not in the document]
Comment (kwannoel) [reply e2e_test/batch/explain.slt:20]: I see, let me use that instead.

File: src/frontend/src/handler/explain.rs [hunk 4]
```diff
@@ -248,6 +252,11 @@ pub async fn handle_explain(
     if analyze {
         bail_not_implemented!(issue = 4856, "explain analyze");
     }
+    if (options.trace || options.explain_type == ExplainType::DistSql)
+        && options.explain_format == ExplainFormat::Json
+    {
+        bail_not_implemented!("explain (trace, distsql, format json) is not supported");
```
Comment (xxchan) [inline src/frontend/src/handler/explain.rs:258]: nit: should it be `not_implemented` or `not_supported`?

besides, the error message looks like `trace && distsql && format json`

**Review by xxchan — APPROVED** [review]

File: src/frontend/src/handler/explain.rs (hunk 4 above)
Comment (kwannoel) [reply src/frontend/src/handler/explain.rs]: Fixed, used the `not supported` error variant.

File: e2e_test/batch/explain.slt (hunk 3 above)
[the remark this answers is not in the document]
Comment (kwannoel) [reply e2e_test/batch/explain.slt]: Fixed

**Pull request merged**